dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,161 @@
+                        )
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+import qualified Data.Map      as M
+import qualified Data.Text     as T
+import qualified Data.Text.IO  as TIO
+import qualified Dhall.Map     as DM (keys)
+
+data FilePos = FP
+    { fpLine :: Int
+    , fpOffset :: Int
+    } deriving (Eq, Ord)
+
+newtype Tags = Tags (M.Map FilePath (M.Map FilePos Tag))
+
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith M.union ts1 ts2)
+
+data Tag = Tag
+    { tagDefinition :: Text
+    , tagName :: Text
+    }
+
+{-| Generate ETags for Dhall expressions
+
+    Arguments:
+
+    * Path to directory or filename that should be indexed
+    * List of suffixes for dhall files [".dhall"] by default
+    * Flag if generate should follow symlinks
+-}
+generate :: Input -> [Text] -> Bool -> IO Text
+generate StandardInput _ _ = showTags . tags "" <$> TIO.getContents
+generate (File p) sxs followSyms = do
+    isD <- doesDirectoryExist p
+    files <- dirToFiles followSyms (map T.unpack (if isD then sxs else [""])) p
+    showTags <$> foldM (\ts f -> do
+                                   t <- tags f <$> handle (\(SomeException _) -> return "")
+                                                          (TIO.readFile f)
+                                   return (ts <> t)) (Tags M.empty) files
+
+{-| Find tags in Text (second argument) and generates list of them
+    To make tags for filenames working in both, emacs and vi, add two initial tags.
+    First for `filename` for vi and second with `/filename` for emacs.
+    Other tags are working for both.
+-}
+tags :: FilePath -> Text -> Tags
+tags f t = Tags (M.singleton f
+                 (M.union initialMap (parse t)))
+    where initialMap = M.fromList [ (FP 1 0, Tag "" (T.pack . takeFileName $ f))
+                                  , (FP 1 1, Tag "" ("/" <> (T.pack . takeFileName) f))
+                                  ]
+
+parse :: Text -> M.Map FilePos Tag
+parse t = M.fromAscList . map (\(FP ln c, Tag _ term) ->
+              let (lsl, l) = mls M.! ln in
+              (FP ln (lsl + c), Tag (T.take c l) term)) . M.assocs $
+          parseExpr e (FP 0 0) M.empty
+    where e = case exprFromText "" t of
+                  Left _ -> None
+                  Right r -> r
+          mls = M.fromList . fst . foldl (\(ls, lsl) (n, l) ->
+                                             let lsl' = lsl + 1 + T.length l in
+                                             ((n, (lsl, l)):ls, lsl'))
+                                        ([], 0) . zip [1..] $ T.lines t
+
+parseExpr :: Expr Src a -> FilePos -> M.Map FilePos Tag -> M.Map FilePos Tag
+parseExpr (Let b e) lpos  mts = parseExpr e lpos (parseBinding b mts)
+parseExpr (Annot e1 e2) lpos mts = parseExpr e1 lpos (parseExpr e2 lpos mts)
+parseExpr (Record mr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $  mr
```

Comment:
I think this would again be clearer with `foldMap`.

Maybe you can even use `subExpressions` instead of writing the traversal yourself, but I'm not sure.

https://github.com/dhall-lang/dhall-haskell/blob/9de7ac51a8e8103e10635b70c647ac002ebebd8d/dhall/src/Dhall/Core.hs#L2190-L2191

Reply:
Made some changes.